source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+    mixin(refract!(answer, "answer").setName("ultimateAnswer").mixture);
+    static assert(
+      __traits(getAttributes, ultimateAnswer) ==
+      __traits(getAttributes, answer));
+  }
+
+  /**
+     Index of function in a set of overloads. Valid only if Function represents
+     a function (not a function pointer), and was refracted from a module.
+   */
+
+  uint index;
+
+  /**
+     Return a new Function object with the index attribute set to value.
+  */
+
+  Function setIndex(uint value)
+  {
+    mixin replaceAttribute!"index";
+    return copy;
+  }
+
+  /**
+     Return a string representing the parameter.
+   */
+
+  /**
+     Return type. Initial value: `std.traits.ReturnType!fun`.
+   */
+
+  string returnType;
+
+  /**
+     Return a new Function object with the returnType attribute set to value.
+  */
+
+  Function setReturnType(string value)
+  {
+    mixin replaceAttribute!"returnType";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure int answer() { return 42; }
+    mixin(
+      refract!(answer, "answer").setName("realAnswer")
+      .setReturnType("real")
+      .mixture);
+    static assert(is(typeof(realAnswer()) == real));
+    static assert(functionAttributes!realAnswer & FunctionAttribute.pure_);
+  }
+
+  /**
+     Function parameters. Initial value: from the refracted function.
+   */
+
+  Parameter[] parameters;
+
+  /**
+     Return a new Function object with the parameters attribute set to value.
+  */
+
+  Function setParameters(immutable(Parameter)[] value)
+  {
+    mixin replaceAttribute!"parameters";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    int answer();
+    mixin(
+      refract!(answer, "answer").setName("answerQuestion")
+      .setParameters([ Parameter().setName("question").setType("string")])
+      .mixture);
+    int control(string);
+    static assert(is(Parameters!answerQuestion == Parameters!control));
+  }
+
+  /**
+     Return a new Function object with a parameter attribute containing the
+     same parameters, with new parameters inserted parameters at the specified
+     index.
+  */
+
+  Function insertParameters(uint index, immutable(Parameter)[] inserted...)
+  {
+    auto value = index == parameters.length ? parameters ~ inserted
+      : index == 0 ? inserted ~ parameters
+      : parameters[0..index] ~ inserted ~ parameters[index..$];
+    mixin replaceAttribute!"parameters";
+    return copy;
+  }
+
+  /**
+     Function body. Initial value: `;`.
+   */
+
+  string body_;
+
+  /**
+     Return a new Function object with the body_ attribute set to value.
+  */
+
+  Function setBody(string value)
+  {
+    mixin replaceAttribute!"body_";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure int answer();
+    mixin(
+      refract!(answer, "answer").setName("theAnswer")
+      .setBody("{ return 42; }")
+      .mixture);
+    static assert(theAnswer() == 42);
+  }
+
+  /**
+     Function attributes.
+     Initial value: `__traits(getAttributes, fun)`.
```

Comment:
Doesn't `__traits(getAttributes, fun)` [return a tuple](https://dlang.org/spec/traits.html#getAttributes)? But this is a `ulong`?

Reply:
> Doesn't `__traits(getAttributes, fun)` [return a tuple](https://dlang.org/spec/traits.html#getAttributes)? But this is a `ulong`?

Ah, correct, `attributes` is the *function* attributes, not UDAs, and they're converted to bit masks